### DISCUSSION TITLE
GET-only route for ServeFile?

### DISCUSSION THREAD

**Discussion by maximecb** [tokio-rs/axum#2788] [Q&A] [answered]
### Summary

I want to have a `GET` route that maps to a static HTML file with a form, along with a `POST` route for the form submission, ideally both with the same name.

The problem I'm running into is that using `route_service` with `ServeFile` seems to occupy both the `GET` and `POST` routes associated with the `/submit` name:

```
 let app = Router::new()
    .route("/", get(get_index))
    .route_service("/submit", ServeFile::new("public/submit.html"))
    .route("/submit", post(submit))
// ...
```

The above produces the following error message:
```
Invalid route "/submit": insertion failed due to conflict with previously registered route: /submit
```

There shouldn't be a conflict though because clearly `ServeFile` can only handle `GET` and not `POST`.  So my question is, is there an easy way to have a route to serve a static file and respond only to the `GET` action, without conflicting with the corresponding `POST` route?

### axum version

0.7.2

**Comment by jplatte** [comment, marked as answer]
Yes, quite easy:

```rust
use axum::routing::{get, get_service};

let app = Router::new()
    .route("/", get(get_index))
    .route(
        "/submit",
        get_service(ServeFile::new("public/submit.html")).post(submit),
    );
```

**Reply by maximecb** [reply]
Thank you that very much does it 🙏 🙂
